README.md
# Funimation Downloader NX

Funimation Downloader NX is capable of downloading videos from the *Funimation* streaming service.

## Legal Warning

This application is not endorsed by or affiliated with *Funimation*. This application enables you to download videos for offline viewing which may be forbidden by law in your country. The usage of this application may also cause a violation of the *Terms of Service* between you and the stream provider. This tool is not responsible for your actions; please make an informed decision before using this application.

## Prerequisites

* NodeJS >= 12.2.0 (https://nodejs.org/)
* NPM >= 5.3.0 (https://www.npmjs.org/)
* ffmpeg >= 4.0.0 (https://www.videohelp.com/software/ffmpeg)
* MKVToolNix >= 20.0.0 (https://www.videohelp.com/software/MKVToolNix)

### Paths Configuration

By default this application uses the following paths to programs (main executables):
* `./modules/mkvtoolnix/mkvmerge`
* `./modules/ffmpeg`

To change these paths you need to edit `config.main.yml` in `./modules/` directory.

### Node Modules

After installing NodeJS with NPM go to directory with `package.json` file and type: `npm i`.
* [check dependencies](https://david-dm.org/seiya-dev/funimation-downloader-nx)

## CLI Options

### Authentication

* `--auth` enter auth mode

### Get Show ID

* `--search <s>` sets the show title for search

### Download Video

* `-s <i> -e <s>` sets the show id and episode ids (comma-separated, hyphen-sequence)
* `-q <i>` sets the video layer quality [1...10] (optional, 0 is max)
* `--alt` alternative episode listing (if available)
* `--sub` switch from English dub to Japanese dub with subtitles
* `--simul` force select simulcast version instead of uncut version
* `-x` select server
* `--nosubs` skip download subtitles for Dub (if available)

### Proxy

* `--proxy <s>` http(s)/socks proxy WHATWG url (ex. https://myproxyhost:1080)
* `--proxy-auth <s>` Colon-separated username and password for proxy
* `--ssp` don't use proxy for stream downloading

### Muxing

`[note] this application mux into mkv by default`
* `--mp4` mux into mp4
* `--mks` add subtitles to mkv or mp4 (if available)

### Filenaming (optional)

* `-a <s>` release group ("Funimation" by default)
* `-t <s>` show title override
* `--ep <s>` episode number override (ignored in batch mode)
* `--suffix <s>` filename suffix override (first "SIZEp" will be replaced with actual video size, "SIZEp" by default)

### Utility

* `--nocleanup` move unnecessary files to trash folder after completion instead of deleting
* `-h`, `--help` show all options

## Filename Template

[`release group`] `title` - `episode` [`suffix`].`extension`

## CLI Examples

* `node funi --search "My Hero"` search "My Hero" in title
* `node funi -s 124389 -e 1,2,3` download episodes 1-3 from show with id 124389
* `node funi -s 124389 -e 1-3,2-7,s1-2` download episodes 1-7 and "S"-episodes 1-2 from show with id 124389


CHANGELOG.md
## Change Log

### 4.4.1 (2019/07/21)
- Fixed proxy for stream download

### 4.4.0 (2019/06/04)
- Added `--novids` option (Thanks to @subdiox)
- Update modules

### 4.3.2 (2019/05/09)
- Code improvements
- Fix `hls-download` error printing

### 4.3.1 (2019/05/09)
- Fix auto detection max quality (Regression in d7d280c)

### 4.3.0 (2019/05/09)
- Better server selection (Closes #42)

### 4.2.1 (2019/05/04)
- Filter duplicate urls for cloudfront.net (Closes #40)

### 4.2.0 (2019/05/02)
- Replace `request` module with `got`
- Changed proxy cli options
- Changed `login` option name to `auth`
- Changed `hls-download` parallel download configuration from 5 parts to 10
- Update modules

### 4.1.0 (2019/04/05)
- CLI options for login moved to CUI
- Removed showing set token at startup

### 4.0.5 (2019/02/09)
- Fix downloading shows with autoselect max quality

### 4.0.4 (2019/01/26)
- Fix search when shows not found
- Update modules

### 4.0.3 (2018/12/06)
- Select only non-encrypted (HLS) streams, encrypted streams is MPEG-DASH

### 4.0.2 (2018/11/25)
- Fix typos and update modules

### 4.0.1 (2018/11/23)
- Code refactoring and small fixes

### 4.0.0 RC 1 (2018/11/17)
- Select range of episodes using hyphen-sequence
- Skip muxing if executables not found
- Fixed typos and duplicate options

### 4.0.0 Beta 2 (2018/11/12)
- Select alternative server
- Updated readme

### 4.0.0 Beta 1 (2018/11/10)
- Rearrange folders structure
- Configuration changed to yaml format
- Muxing changed to MKV by default
- tsMuxeR+mp4box replaced with FFMPEG
- Updated commands help and readme
- Fixed typos and duplicate options
- `ttml2srt` moved to separate module
- Drop `m3u8-stream-list` module
- Code improvements

### 3.2.8 (2018/06/16)
- Fix video request when token not specified

### 3.2.7 (2018/06/15)
- Update modules

### 3.2.6 (2018/02/18)
- Fix commands help

### 3.2.5 (2018/02/12)
- Fixes and update modules

### 3.2.4 (2018/02/01)
- Update modules

### 3.2.3 (2018/01/31)
- Rearrange folders structure

### 3.2.2 (2018/01/16)
- Update modules

### 3.2.1 (2018/01/16)
- Update modules
- Small fixes

### 3.2.0 (2018/01/16)
- `hls-download` module moved to independent module
- Auth for socks proxy

### 3.1.0 (2017/12/30)
- Convert DXFP (TTML) subtitles to SRT format

### 3.0.1 (2017/12/05)
- Check subtitles availability
- Download subtitles in SRT format instead of VTT
- Extended hls download progress info

### 3.0.0 Beta 3 (2017/12/03)
- Restored MKV and MP4 muxing
- Convert VTT subtitles to SRT format

### 3.0.0 Beta 2 (2017/10/18)
- Fix video downloading

### 3.0.0 Beta 1 (2017/10/17)
- Major code changes and improvements
- Drop Streamlink and added own module for hls download

### 2.5.0 (2017/09/04)
- `nosubs` option
- Request video with app api

### 2.4.1 (2017/09/02)
- Fixed typo in package.json
- Fix #11: URL for getting video stream url was changed

### 2.4.0 (2017/07/04)
- IPv4 Socks5 proxy support

### 2.3.3 (2017/06/19)
- Removed forgotten debug code

### 2.3.2 (2017/06/19)
- Fix #5: Script fails to multiplex unique file names

### 2.3.1 (2017/04/29)
- Code improvements

### 2.3.0 (2017/04/27)
- Code improvements

### 2.2.5 (2017/04/17)
- Minor code improvements and fixes

### 2.1.4 (2017/04/10)
- Minor changes

### 2.1.3 (2017/04/10)
- Minor changes and fixes

### 2.1.2 (2017/04/10)
- Fix config path

### 2.1.1 (2017/04/10)
- Minor text changes
- Fix config
- Minor changes

### 2.1.0 (2017/04/10)
- First stable release

### 2.0.0 Beta (lost in time)
- First public release

scripts/module.ttml-parser.js
function ttml2srt(data) {
	let f = data.match(/ttp:frameRate\s*=\s*"(.*?)"/);
	let frameRate = f ? parseInt(f[1]) : 25;
	let reStr = '<p style.*?begin="([^"]*)" end="([^"]*)".*?>(.*?)</p>';
	let gre = new RegExp(reStr, 'g');
	let re = new RegExp(reStr);
	let res = '';
	let str_id = 0;
	for (let x of data.match(gre)) {
		let m = x.match(re);
		if (m) {
			let begin = formatSrtTime(m[1], frameRate);
			let end = formatSrtTime(m[2], frameRate);
			let text = m[3]
				.replace(/(<br.*?>)+/g, '\n')
				.replace(/<(\S*?) (.*?)>(.*?)<\/.*?>/g, fontRepl);
			if(text.trim() !== ''){
				str_id++;
				res += `${str_id}\n${begin} --> ${end}\n${text}\n\n`;
			}
		}
	}
	return res;
}
function formatSrtTime(time, frameRate) {
	let t = time.match(/(.*):([^:]*)$/);
	let ms = Math.floor(parseInt(t[2]) * 1000 / frameRate).toString();
	return t[1] + ',' + ms.padStart(3, '0');
}
function fontRepl(str, tag, attrs, txt) {
	if (tag != 'span') {
		return txt;
	}
	let at = attrs.replace(/\s*=\s*/g, '=').split(' ').filter(x => x.trim());
	for (let a of at) {
		let ax = a.match(/tts:color="(.*?)"/);
		if (ax) {
			txt = `<font color="${ax[1]}">${txt}</font>`;
			continue;
		}
		switch (a) {
			case 'tts:fontStyle="italic"':
				txt = `<i>${txt}</i>`;
				break;
			case 'tts:textDecoration="underline"':
				txt = `<u>${txt}</u>`;
				break;
			case 'tts:fontWeight="bold"':
				txt = `<b>${txt}</b>`;
				break;
		}
	}
	return txt;
}

module.exports = (ttml) => {
	return `\uFEFF${ttml2srt(ttml).replace(/\n/g,'\r\n')}`;
};


scripts/module.hls-download.js
const agent = require('socks5-https-client/lib/Agent');
const crypto = require('crypto');
const request = require('request');
const shlp = require('sei-helper');
const fs = require('fs');
let date_start;

// async
function getData(method, url, proxy) {
	// base options
	let options = {
		headers: {
			'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; WOW64; rv:55.0) Gecko/20100101 Firefox/55.0'
		}
	};
	if (proxy && proxy.type === 'socks') {
		options.agentClass = agent;
		let agentOptions = {
			socksHost: proxy.ip.split(':')[0],
			socksPort: proxy.ip.split(':')[1]
		};
		options.agentOptions = agentOptions;
		options.timeout = 10000;
	}
	else if (proxy && proxy.type === 'http') {
		options.proxy = 'http://' + proxy.ip;
		options.timeout = 10000;
	}
	// request parameters
	options.url = url;
	options.encoding = null;
	// do request
	return new Promise((resolve, reject) => {
		request[method](options, (err, res) => {
			if (err) return reject(err);
			if (res.statusCode != 200) {
				return reject(new Error(`Response code: ${res.statusCode}. Body: ${res.body}`));
			}
			resolve(res);
		});
	});
}

function getURI(baseurl, uri) {
	const httpURI = /^https{0,1}:/.test(uri);
	if (!baseurl && !httpURI) {
		throw new Error('no base and not http uri');
	} else if (httpURI) {
		return uri;
	}

	return baseurl + uri;
}

async function dlparts(m3u8json, fn, baseurl, proxy) {
	let keys = {}
	// delete file if exists
	if(fs.existsSync(`${fn}.ts`)){
		fs.unlinkSync(`${fn}.ts`);
	}
	// dl parts
	const pcount = 10;
	for (let p = 0; p < m3u8json.segments.length / pcount; p++) {
		let offset = p * pcount;
		let prq = new Map();
		for (let px = offset; px < offset + pcount && px < m3u8json.segments.length; px++) {
			prq.set(px, dlpart(m3u8json, fn, px, baseurl, keys, proxy));
		}
		let res = [];
		for (let x = 5; x--;) {
			for (let i = prq.size; i--;) {
				try {
					let r = await Promise.race(prq.values());
					prq.delete(r.p);
					res[r.p - offset] = r.dec;
				} catch (error) {
					console.log(`[ERROR] Part ${error.p} download error: ${error.message}${x > 0 ? ', retry' : ', FAIL'}`);
					prq.set(error.p, dlpart(m3u8json, fn, error.p, baseurl, keys, proxy));
				}
			}
		}
		if (prq.size > 0) {
			throw new Error(`${prq.size} parts not downloaded`);
		}
		let dled = offset + 10;
		
		getDLedInfo((dled<m3u8json.segments.length?dled:m3u8json.segments.length),m3u8json.segments.length);
		
		for (let r of res) {
			fs.writeFileSync(`${fn}.ts`, r, { flag: 'a' });
		}
	}
}

function getDLedInfo(dled,total){
	const date_elapsed = Date.now() - date_start;
	const percentFxd = (dled / total * 100).toFixed();
	const percent = percentFxd < 100 ? percentFxd : 99;
	const time = shlp.htime(((parseInt(date_elapsed * (total / dled - 1))) / 1000).toFixed());
	console.log(`[INFO] ${dled} parts of ${total} downloaded [${percent}%] (${time})`);
}

async function getDecipher(pd, keys, baseurl, proxy) {
	const kURI = getURI(baseurl, pd.key.uri);
	if (!keys[kURI]) {
		const rkey = await getData('get', kURI, proxy);
		if (!rkey || !rkey.body) {
			throw new Error('key get error');
		}
		keys[kURI] = rkey.body;
	}
	// get ivs
	let iv = Buffer.alloc(16);
	let ivs = pd.key.iv;
	for (i in ivs) {
		iv.writeUInt32BE(ivs[i], i * 4);
	}
	return crypto.createDecipheriv('aes-128-cbc', keys[kURI], iv);
}

async function dlpart(m3u8json, fn, p, baseurl, keys, proxy) {
	// console.log(`download segment ${p+1}`);
	let pd = m3u8json.segments[p];
	let decipher, part;
	try {
		decipher = await getDecipher(pd, keys, baseurl, proxy);
		part = await getData('get', getURI(baseurl, pd.uri), proxy);
	} catch (error) {
		error.p = p;
		throw error;
	}
	let dec = decipher.update(part.body);
	dec = Buffer.concat([dec, decipher.final()]);
	let part_num = p + 1;
	let part_num_lng = part_num.toString().length;
	// console.log('Part', '0'.repeat(ttlLng-part_num_lng)+part_num, 'of', m3u8json.segments.length, 'downloaded');
	return { dec, p }
}

let ttlLng = 0;
module.exports = async (m3u8json, fn, baseurl, proxy) => {
	// console.log({m3u8json, fn, baseurl});
	console.log('[INFO] Starting downloading ts...')
	ttlLng = m3u8json.segments.length.toString().length;
	let res = { "ok": true };
	try {
		date_start = Date.now();
		await dlparts(m3u8json, fn, baseurl, proxy);
	} catch (error) {
		res = { "ok": false, "err": error };
	}
	return res;
};


scripts/funidl.js
// package json
const packageJson = require('./package.json');

// program name
console.log('\n=== Funimation Downloader NX '+packageJson.version+' ===\n');
const api_host = 'https://prod-api-funimationnow.dadcdigital.com/api';

// modules build-in
const { chdir } = require('process');
const path = require('path');
const fs = require('fs');

// modules extra
const shlp = require('sei-helper');
const yargs = require('yargs');
const request = require('request');
const agent = require('socks5-https-client/lib/Agent');

// folders
const configDir  = path.join(__dirname,'/config/');
const configBase = path.join(__dirname,'/base/');
let bin, workDir = {};

// check folders
if(fs.existsSync(configBase)){
	bin             = require(path.join(configBase,'/config.bin.js'));
	workDir         = require(path.join(configBase,'/config.dir.js'));
}
else{
	bin             = require(path.join(configDir,'/config.bin.js'));
	workDir.content = path.join(__dirname,'/../videos/');
	workDir.trash   = path.join(__dirname,'/../videos/_trash/');
}

// auth check
let token = false;
const cfgFilename = configDir + '/funi_auth.json';
if(fs.existsSync(cfgFilename)){
	token = require(cfgFilename).token;
}

// cli
let argv = yargs
	.wrap(Math.min(100))
	.usage('Usage: $0 [options]')
	
	// login
	.describe('mail','Your email')
	.describe('pass','Your password')
	
	// params
	.describe('s','Set show id')
	.describe('alt','Alternative episode listing (if available)')
	.boolean('alt')
	
	.describe('sel','Select episode')
	.describe('sub','Subtitles mode (Dub mode by default)')
	.boolean('sub')
	
	.describe('q','Video quality')
	.choices('q', ['234p','270p','288p','360p','480p','540p','720p','1080p'])
	.default('q','720p')
	
	.describe('a','Release group')
	.default('a','Funimation')
	.describe('t','Filename: series title override')
	.describe('ep','Filename: episode number override')
	.describe('suffix','Filename: filename suffix override (first "SIZEp" will be raplaced with actual video size)')
	.default('suffix','SIZEp')
	
	.describe('nosubs','skip download subtitles for Dub (if available)')
	.boolean('nosubs')
	
	.describe('mkv','Mux into mkv')
	.boolean('mkv')
	.describe('mks','Add subtitles to mkv (if available)')
	.boolean('mks')
	
	// login
	.describe('mail','Your email')
	.describe('pass','Your password')
	
	// proxy
	.describe('socks','Set ipv4 socks5 proxy')
	.describe('proxy','Set ipv4 http(s) proxy')
	
	// help
	.describe('h','Show this help')
	.alias('h','help')
	.boolean('h')
	
	.argv;

// check page
if(!isNaN(parseInt(argv.p, 10)) && parseInt(argv.p, 10) > 0){
	argv.p = parseInt(argv.p, 10);
}
else{
	argv.p = 1;
}

// check proxy
if(argv.socks){
	if(!shlp.validateIpAndPort(argv.socks)){
		console.log('Error: not ipv4 socks5 proxy. Skipping...\n');
		argv.socks = undefined;
	}
}
else if(argv.proxy){
	if(!shlp.validateIpAndPort(argv.proxy)){
		console.log('Error: not ipv4 http(s) proxy. Skipping...\n');
		argv.proxy = undefined;
	}
}

// fn variables
let fnTitle = '',
	fnEpNum = '',
	fnSuffix = '',
	fnOutput = '',
	fnOutput_bak = '',
	tsDlPath = false,
	stDlPath = false;

// select mode
if(argv.mail && argv.pass){
	auth();
}
else if(argv.search || argv.s && !isNaN(parseInt(argv.s,10)) && parseInt(argv.s,10) > 0){
	if(argv.search){
		searchShow();
	}
	else{
		getShow();
	}
}
else{
	console.log(yargs.help());
	process.exit();
}

async function auth(){
	let authData;
	try{
		authData = await getData(api_host+'/auth/login/',false,true,false,true);
		checkResp(authData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	authData = JSON.parse(authData.body);
	if(authData.token){
		console.log('[INFO] Authentication success, your token:',authData.token.slice(0,7)+'*'.repeat(33),'\n');
		fs.writeFileSync(cfgFilename,JSON.stringify({"token":authData.token},null,'\t'));
	}
	else{
		console.log('[ERROR]',authData.error,'\n');
		process.exit(1);
	}
}

async function searchShow(){
	let searchData;
	try{
		let qs = {unique:true,limit:100,q:argv.search,offset:(argv.p-1)*1000};
		searchData = await getData(api_host+'/source/funimation/search/auto/',qs,true,true);
		checkResp(searchData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	
	searchData = JSON.parse(searchData.body);
	if(searchData.items.hits){
		let shows = searchData.items.hits;
		console.log('[INFO] Search Results:');
		for(let ssn in shows){
			console.log('[#'+shows[ssn].id+'] '+shows[ssn].title+(shows[ssn].tx_date?' ('+shows[ssn].tx_date+')':''));
		}
	}
	console.log('[INFO] Total shows found:',searchData.count,'\n');
}

async function getShow(){
	// show main data
	let showData;
	try{
		showData = await getData(api_host+'/source/catalog/title/'+parseInt(argv.s,10),false,true,true);
		checkResp(showData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	// check errors
	showData = JSON.parse(showData.body);
	if(showData.status){
		console.log('[ERROR] Error #'+showData.status+':',showData.data.errors[0].detail,'\n');
		process.exit(1);
	}
	else if(!showData.items || showData.items.length<1){
		console.log('[ERROR] Show not found\n');
	}
	showData = showData.items[0];
	console.log('[#'+showData.id+'] '+showData.title+' ('+showData.releaseYear+')');
	// show episodes
	let episodesData;
	try{
		let qs = {limit:-1,sort:'order',sort_direction:'ASC',title_id:parseInt(argv.s,10)};
		if(argv.alt){ qs.language = 'English'; }
		episodesData = await getData(api_host+'/funimation/episodes/',qs,true,true);
		checkResp(episodesData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	// parse episodes list
	let eps = JSON.parse(episodesData.body).items,
		fnSlug = false;
	for(let e in eps){
		let showStrId = eps[e].ids.externalShowId;
		let epStrId = eps[e].ids.externalEpisodeId.replace(new RegExp('^'+showStrId),'');
		// select
		if(epStrId == argv.sel){
			fnSlug = {title:eps[e].item.titleSlug,episode:eps[e].item.episodeSlug};
		}
		// console vars
		let tx_snum = eps[e].item.seasonNum==1?'':' S'+eps[e].item.seasonNum;
		let tx_type = eps[e].mediaCategory != 'episode' ? eps[e].mediaCategory : '';
		let tx_enum = eps[e].item.episodeNum !== '' ? '#' + (eps[e].item.episodeNum < 10 ? '0'+eps[e].item.episodeNum : eps[e].item.episodeNum) : '#'+eps[e].item.episodeId;
		let qua_str = eps[e].quality.height ? eps[e].quality.quality +''+ eps[e].quality.height : 'UNK';
		let aud_str = eps[e].audio.length > 0 ? ', '+eps[e].audio.join(', ') : '';
		let rtm_str = eps[e].item.runtime !== '' ? eps[e].item.runtime : '??:??';
		// console string
		let episodeIdStr = epStrId;
		let conOut  = '['+episodeIdStr+'] ';
			conOut += eps[e].item.titleName+tx_snum + ' - ' +tx_type+tx_enum+ ' ' +eps[e].item.episodeName+ ' ';
			conOut += '('+rtm_str+') ['+qua_str+aud_str+ ']';
			conOut += epStrId == argv.sel ? ' (selected)' : '';
			conOut += eps.length-1 == e ? '\n' : '';
		console.log(conOut);
	}
	if(!fnSlug){
		process.exit();
	}
	// parse episode data
	let episodeData;
	try{
		episodeData = await getData(api_host+'/source/catalog/episode/'+fnSlug.title+'/'+fnSlug.episode+'/',false,true,true);
		checkResp(episodeData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	let ep = JSON.parse(episodeData.body).items[0], streamId = 0;
	// build fn
	fnTitle = argv.t ? argv.t : ep.parent.title;
	ep.number = isNaN(ep.number) ? ep.number : ( parseInt(ep.number, 10) < 10 ? '0' + ep.number : ep.number );
	if(ep.mediaCategory != 'Episode'){
		ep.number = ep.number !== '' ? ep.mediaCategory+ep.number : ep.mediaCategory+'#'+ep.id;
	}
	fnEpNum = argv.ep ? ( parseInt(argv.ep, 10) < 10 ? '0' + argv.ep : argv.ep ) : ep.number;
	fnSuffix = argv.suffix.replace('SIZEp',argv.q);
	fnOutput = shlp.cleanupFilename('['+argv.a+'] ' + fnTitle + ' - ' + fnEpNum + ' ['+ fnSuffix +']');
	// end
	console.log('[INFO] Output filename: '+fnOutput,'\n\n[INFO] Available audio tracks:');
	for(let m in ep.media){
		let selected = false;
		if(ep.media[m].mediaType=='experience'){
			let media_id = ep.media[m].id;
			let dub_type = ep.media[m].title.split('_')[1];
			if(dub_type == 'Japanese' && argv.sub){
				streamId = ep.media[m].id;
				stDlPath = getSubsUrl(ep.media[m].mediaChildren);
				selected = true;
			}
			else if(dub_type == 'English' && !argv.sub){
				streamId = ep.media[m].id;
				stDlPath = getSubsUrl(ep.media[m].mediaChildren);
				selected = true;
			}
			console.log('[#'+media_id+'] '+dub_type+(selected?' (selected)':''));
		}
	}
	if(streamId<1){
		console.log('\n[ERROR] Dub not selected\n');
		process.exit();
	}
	// get stream url
	let streamData;
	try{
		streamData = await getData(api_host+'/source/catalog/video/'+streamId+'/signed',{"dinstid":"uuid"},true,true);
		checkResp(streamData.body);
	}
	catch(error){
		console.log(error,'\n');
		process.exit(1);
	}
	streamData = JSON.parse(streamData.body);
	if(streamData.errors){
		console.log('\n[ERROR] Error #'+streamData.errors[0].code+':',streamData.errors[0].detail,'\n');
		process.exit(1);
	}
	else{
		for(let u in streamData.items){
			if(streamData.items[u].videoType == 'm3u8'){
				tsDlPath = streamData.items[u].src;
				break;
			}
		}
	}
	if(!tsDlPath){
		console.log('\n[ERROR] Unknown error\n');
		process.exit(1);
	}
	downloadStreams();
}

function getSubsUrl(m){
	if(argv.nosubs && !argv.sub){
		return false;
	}
	for(let i in m){
		let fpp = m[i].filePath.split('.');
		let fpe = fpp[fpp.length-1];
		if(fpe == 'vtt'){
			return m[i].filePath;
		}
	}
	return false;
}

async function downloadStreams(){
	// to work dir
	chdir(workDir.content);
	// download video
	let speedupSegments = '--hls-segment-attempts 10 --hls-segment-threads 10 --hls-segment-timeout 60';
	if(!argv.nots){
		shlp.exec('streamlink','"'+path.normalize(bin.streamlink)+'"','"hlsvariant://'+tsDlPath+'" '+argv.q+' '+speedupSegments+' -o "'+fnOutput+'.ts"',true);
	}
	// download subtitles
	if(stDlPath){
		console.log('\n[INFO] Downloading subtitles...');
		let subsSrc = await getData(stDlPath);
		fs.writeFileSync(fnOutput+'.vtt',subsSrc.body);
		console.log('[INFO] Downloaded!');
	}
	// select muxer
	if(argv.mkv){
		// mux to mkv
		let mkvmux  = '-o "'+fnOutput+'.mkv" --disable-track-statistics-tags --engage no_variable_data ';
			mkvmux += '--track-name "0:['+argv.a+']" --language "1:'+(argv.sub?'jpn':'eng')+'" --video-tracks 0 --audio-tracks 1 --no-subtitles --no-attachments ';
			mkvmux += '"'+fnOutput+'.ts" ';
			if(argv.mks && stDlPath){
				mkvmux += '--language 0:eng "'+fnOutput+'.vtt" ';
			}
		shlp.exec('mkvmerge','"'+path.normalize(bin.mkvmerge)+'"',mkvmux,true);
		if(!argv.nocleanup){
			fs.renameSync(fnOutput+'.ts', workDir.trash+'/'+fnOutput+'.ts');
		}
	}
	else{
		// check filename for ts muxer
		fnOutput_bak = fnOutput;
		if(fnOutput_bak.indexOf('.')>-1){
			fnOutput = fnOutput.replace(/\./g,'_');
			fs.renameSync(fnOutput_bak+'.ts', fnOutput+'.ts');
		}
		// Get stream data
		let metaData = require('child_process').execSync('"'+path.normalize(bin.tsmuxer)+'" "'+fnOutput+'.ts"');
		let metaDataRe = /Track ID:\s*(\d+)[\s\S]*?Stream ID:\s*([\S]*)[\s\S]*?Frame rate:\s*([\S]*)[\s\S]*?Track ID:\s*(\d+)[\s\S]*?Stream ID:\s*([\S]*)[\s\S]*?Stream delay:\s*([\S]*)/;
		let metaArgs = metaData.toString().match(metaDataRe);
		// demux streams
		let ts2meta  = 'MUXOPT --no-pcr-on-video-pid --new-audio-pes --demux --vbr  --vbv-len=500\n';
			ts2meta += metaArgs[2]+', "'+path.normalize(workDir.content+'/'+fnOutput+'.ts')+'", insertSEI, contSPS, track='+metaArgs[1]+'\n';
			ts2meta += metaArgs[5]+', "'+path.normalize(workDir.content+'/'+fnOutput+'.ts')+'", timeshift='+metaArgs[6]+'ms, track='+metaArgs[4];
		fs.writeFileSync(fnOutput_bak+'.meta',ts2meta);
		shlp.exec('tsmuxer','"'+path.normalize(bin.tsmuxer)+'"','"'+fnOutput_bak+'.meta" "'+path.normalize(workDir.content)+'"',true);
		if(fnOutput_bak.indexOf('.')>-1){
			fs.renameSync(fnOutput+'.track_'+metaArgs[1]+'.264',fnOutput_bak+'.264');
			fs.renameSync(fnOutput+'.track_'+metaArgs[4]+'.aac',fnOutput_bak+'.aac');
			fs.renameSync(fnOutput+'.ts', fnOutput_bak+'.ts');
			fnOutput = fnOutput_bak;
		}
		else{
			fs.renameSync(fnOutput+'.track_'+metaArgs[1]+'.264',fnOutput+'.264');
			fs.renameSync(fnOutput+'.track_'+metaArgs[4]+'.aac',fnOutput+'.aac');
		}
		// mux to mp4
		let mp4mux  = '-add "'+fnOutput+'.264#video:name=['+argv.a+']" ';
			mp4mux += '-add "'+fnOutput+'.aac#audio:lang='+(argv.sub?'jpn':'eng')+':name=" ';
			mp4mux += '-new "'+fnOutput+'.mp4" ';
		shlp.exec('mp4box','"'+path.normalize(bin.mp4box)+'"',mp4mux,true);
		// cleanup
		if(!argv.nocleanup){
			fs.unlinkSync(fnOutput+'.meta');
			fs.renameSync(fnOutput+'.ts', workDir.trash+'/'+fnOutput+'.ts');
			fs.renameSync(fnOutput+'.264', workDir.trash+'/'+fnOutput+'.264');
			fs.renameSync(fnOutput+'.aac', workDir.trash+'/'+fnOutput+'.aac');
		}
	}
	console.log('\n[INFO] Done!\n');
}

// check response
function checkResp(r){
	if(r.match(/<!doctype html>/) || r.match(/<html/)){
		console.log('[ERROR] unknown error, body:\n',r,'\n');
		process.exit(1);
	}
}

// get data fro url
function getData(url,qs,proxy,useToken,auth){
	let options = {};
	// request parameters
	options.url = url;
	if(qs){
		options.qs = qs;
	}
	if(auth){
		options.method = 'POST';
		options.formData = {
			username: argv.mail,
			password: argv.pass
		};
	}
	if(useToken && token){
		options.headers = {
			Authorization: 'Token '+token
		};
		if(options.qs && options.qs.dinstid){
			options.headers.devicetype = 'Android Phone';
			delete options.qs;
		}
	}
	if(proxy && argv.socks){
		options.agentClass = agent;
		let agentOptions = {
			socksHost: argv.socks.split(':')[0],
			socksPort: argv.socks.split(':')[1]
		};
		options.agentOptions = agentOptions;
		options.timeout = 10000;
	}
	else if(proxy && argv.proxy){
		options.proxy = 'http://'+argv.proxy;
		options.timeout = 10000;
	}
	// do request
	return new Promise((resolve, reject) => {
		request(options, (err, resp) => {
			if (err) return reject(err);
			if (resp.statusCode != 200 && resp.statusCode != 403) {
				return reject(new Error(`\nStatus: ${resp.statusCode}`));
			}
			resolve(resp);
		});
	});
}
